misc: add Containerfile

mvo5 opened the pull request:
This PR adds Containerfile integration, with that a bib style deploy/run is possible. The Containerfile is pretty nasty right now, especially around the "getting supported repositories" part. This needs to get fixed either in composer or images IMHO (or via go:embed). But with that we can give it to our early users and let them play with building images.

It also contains automatic container building as part of the GH action (thanks @supakeen!) and once we have it we can update the README to point to the auto-generated containers for even easier testing/using.

As is "list-images" and "manifest" will work but once #8 is merged the below works (on my local system, we want tests here :)

Run build via container, bib-style:
```console
$ sudo podman build . -t test-icbli
$ sudo podman run --privileged \
   -v ./output:/output \
   test-ibcli:latest \
   build \
   --distro fedora-41 \
   minimal-raw
```

supakeen commented:
I've been rude and committed CI for the container build, we might need to change organisation or repository permissions for it to upload.

supakeen reviewed (approved):
Looks good (though I guess I wrote part of it so maybe wait for a second review).

ondrejbudai commented on Containerfile:
```diff
@@ -0,0 +1,59 @@
+
+FROM registry.fedoraproject.org/fedora:41
+# Fast-track osbuild so we don't depend on the "slow" Fedora release process to implement new features in bib
+RUN dnf install -y dnf-plugins-core
```
Can we put these in a single RUN command, and clean the dnf cache afterwards, to save some disk space and reduce the number of layers?

```
RUN dnf install -y dnf-plugins-core \
  && dnf copr enable -y @osbuild/osbuild \
  && dnf install -y libxcrypt-compat wget osbuild osbuild-ostree osbuild-depsolve-dnf osbuild-lvm2 \
  && dnf clean all
```

supakeen replied:
I've addressed this in a separate commit.

ondrejbudai commented on README.md:
```diff
@@ -2,6 +2,19 @@
 
+## Run via container
+
+TODO: build container automatically via GH action and put into ghcr.io
```
Leftover TODO

supakeen replied:
I've addressed this in a separate commit.

ondrejbudai commented on entrypoint.sh:
```diff
@@ -0,0 +1,24 @@
+mkdir /run/osbuild-store
+
+mount -t tmpfs tmpfs /run/osbuild
+mount -t tmpfs tmpfs /run/osbuild-store
```
I think this is currently unused, and I think we should actually use a "real" filesystem for the osbuild store, because tmpfs might not be big enough for image building.

mvo5 replied:
Thanks, indeed, I removed this and followed our pattern from bib and added `--store` and use that in the entrypoint.

ondrejbudai commented on Containerfile:
```diff
@@ -0,0 +1,59 @@
+VOLUME /output
+WORKDIR /output
+VOLUME /store
+VOLUME /rpmmd
```
I think this is currently unused.

mvo5 replied:
Indeed, I removed it (for now, I think we want to add this later again once we grow a --rpmmd like flag).

ondrejbudai commented on Containerfile:
```diff
@@ -0,0 +1,59 @@
+ENTRYPOINT ["/entrypoint.sh"]
+VOLUME /output
+WORKDIR /output
+VOLUME /store
```
This is currently unused, and should be used (osbuild's store cannot be on on overlayfs). How can we pass the `/store` path to image builder cli to utilize it for the osbuild's store?

ondrejbudai commented on .github/workflows/container.yaml:
```diff
@@ -0,0 +1,46 @@
+on:
+  push:
+    branches: [ "main" ]
+  pull_request:
```
Also build in a merge queue?

supakeen commented:
I've made 3 separate commits to take care of the low hanging fruit things, feel free to reorganize/rebase how you see fit @mvo5 

(I don't want to force push to your branch at all).

The other remarks will take a bit more thinking.

mvo5 commented:
> I've made 3 separate commits to take care of the low hanging fruit things, feel free to reorganize/rebase how you see fit @mvo5
> 
> (I don't want to force push to your branch at all).
> 
> The other remarks will take a bit more thinking.

Thank you! This is really appreciated! I squashed them into their relevant places (hopefully :) and also addressed the points from @ondrejbudai - I think this is ready for a re-review. I can split out further if needed but it seems it's not toooo big .

achilleas-k reviewed (approved):
LGTM.

achilleas-k commented:
One thought about the new `--store` flag (to consider in a follow-up):
Would it make sense to separate osbuild flags into their own group?  Could we have an `--osbuild-opts` flag for example that can take options to be passed down to osbuild?

So we could do:
```
image-builder-cli ... --osbuild-opts="store=/data/osbuild-store,checkpoint=build,checkpoint=os,cache-max-size='2 GiB'" ...
```

It's not the prettiest but it saves us from having to support every osbuild flag.  OTOH, it also forces users to look up osbuild flags if they want to discover everything that's supported.

If we don't like this, I think eventually we should support every option that osbuild supports (except maybe the `--monitor` option).

ondrejbudai commented on .github/workflows/go.yml:
```diff
@@ -13,6 +13,7 @@ jobs:
 
   build:
     runs-on: ubuntu-latest
+    uses:  ./.github/workflows/testdeps.yml
```
😍

ondrejbudai commented on test/containerbuild.py:
```diff
@@ -0,0 +1,23 @@
+import os
```
I think we just need `subprocess` and `pytest` here.

mvo5 replied:
Yeah, silly me - and a linter :)

ondrejbudai commented on test/test_container.py:
```diff
@@ -0,0 +1,27 @@
+import os
```
Would you be willing to use Go for integration tests? Is there any reason to actually use Python?

I can tell you why I prefer Go:
- developers don't need to switch their language mindset when writing integration tests
- we don't need an extra linter and formatter
- sharing code between integration tests and the SUT
- pytest fixtures are cool... until they start being overwhelming. I think I have a preference for the lightweight nature of Go.

mvo5 replied:
Sure, I will give this a "go" (see what I did here ;)

ondrejbudai commented on Containerfile:
```diff
@@ -0,0 +1,59 @@
+VOLUME /output
+WORKDIR /output
+# XXX: add "store" flag like bib
+VOLUME /store
```
I had an idea: image-builder-cli is supposed to be an RPM in the future. So this path should be FHS-compliant, and ideally it should be the same between the container image and the RPM.

Something like `/var/cache/image-builder/{store,dnf-cache}`?

Note that I have no concerns about `/output`, this is perfectly valid to be container-specific.

mvo5 replied:
Love it! will do in a followup

ondrejbudai commented on test/test_container.py:
```diff
@@ -0,0 +1,27 @@
+        build_container,
+        "build",
+        "minimal-raw",
+        "--distro", "fedora-41"
```
Can we do something quick and dirty to pave over the RPM downloading issue that our stack is suffering from? Override repos with rpmrepo.osbuild.org for the test?

mvo5 replied:
I will try centos-9 instead, from my experience with bib it seems the centos mirrors are generally a bit more stable.

ondrejbudai commented on .github/workflows/pytest.yml:
```diff
@@ -0,0 +1,24 @@
+  push:
+    branches: [ "main" ]
+  pull_request:
+    branches: [ "main" ]
```
no merge queue here
